### PR TITLE
Apply Fork Choice at End of Ancestor Block Fetching

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #1586 

---

# Description

**Write why you are making the changes in this pull request**

Right now, we have a lot of calls to apply fork choice and a high possibility of creating a ton of goroutines in our receive block function in regular sync. The way it works is as follows:

See a block, do we have its parent?
    - **No**: store it in a map of pending child blocks, and issue a call for a parent request.
    - **Yes**: Does the block have pending children we put in the prior map? Then call receiveBlock recursively and apply fork choice on the block and all its children. If the block does not have pending children at the end of the recursive call, end.

This creates a lot of calls to ApplyForkChoice and also a lot of calls to Jaeger spans, which do not end until the child span returns, which can lead to tons of goroutines. Applying fork choice at every step instead of the very last block might be leading to dozens of reorgs and a broken chain upon a restart.

**Write a summary of the changes you are making**

This abstracts the recursive logic into its own function that does not spawns tons of jaeger calls, and applies fork choice only at the very end of the receiveBlock function call once all recursive calls have ended. This should help our runtime and hopefully solve the many reorgs problem.

**Link anything that would be helpful or relevant to the reviewers**

### CODE REVIEW

File: beacon-chain/sync/receive_block.go [hunk 1]
```diff
@@ -65,21 +72,25 @@ func (rs *RegularSync) receiveBlock(msg p2p.Message) error {
...
 	// and call receiveBlock recursively. The recursive function call will stop once
 	// the block we process no longer has children.
 	if child, ok := rs.hasChild(blockRoot); ok {
 		// We clear the block root from the pending processing map.
 		rs.clearPendingBlock(blockRoot)
-		return rs.receiveBlock(child)
+		return rs.processBlockAndFetchAncestors(ctx, child)
 	}
-	return nil
+	return rs.chainService.ApplyForkChoiceRule(ctx, block, beaconState)
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Shouldn't this go in receive block? `processBlockAndFetchAncestors` is still called recursively so you will be calling `ApplyForkChoiceRule` still n times.

File: beacon-chain/sync/receive_block.go [hunk 2]
```diff
@@ -65,21 +72,25 @@ func (rs *RegularSync) receiveBlock(msg p2p.Message) error {
 		return nil
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: No warning or anything? Can we log something here?

File: beacon-chain/sync/receive_block.go [hunk 3]
```diff
@@ -88,36 +99,30 @@ func (rs *RegularSync) validateAndProcessBlock(ctx context.Context, blockMsg p2p
...
 	if block.Slot < beaconState.FinalizedEpoch*params.BeaconConfig().SlotsPerEpoch {
 		log.Debug("Discarding received block with a slot number smaller than the last finalized slot")
-		return [32]byte{}, false, nil
+		return nil, nil, false, err
 	}
 
 	// We check if we have the block's parents saved locally.
 	parentRoot := bytesutil.ToBytes32(block.ParentRootHash32)
 	hasParent := rs.db.HasBlock(parentRoot)
-	span.AddAttributes(trace.BoolAttribute("hasParent", hasParent))
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Revert this change

File: beacon-chain/sync/receive_block.go [hunk 4]
```diff
@@ -65,21 +72,25 @@ func (rs *RegularSync) receiveBlock(msg p2p.Message) error {
...
 		rs.clearPendingBlock(blockRoot)
-		return rs.receiveBlock(child)
+		return rs.processBlockAndFetchAncestors(ctx, child)
 	}
-	return nil
+	return rs.chainService.ApplyForkChoiceRule(ctx, block, beaconState)
 }
 
-func (rs *RegularSync) validateAndProcessBlock(ctx context.Context, blockMsg p2p.Message) ([32]byte, bool, error) {
-	ctx, span := trace.StartSpan(ctx, "beacon-chain.sync.validateAndProcessBlock")
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Revert this change (add back the span)

File: beacon-chain/sync/receive_block.go [hunk 5]
```diff
@@ -88,36 +99,30 @@ func (rs *RegularSync) validateAndProcessBlock(ctx context.Context, blockMsg p2p
 	blockRoot, err := hashutil.HashBeaconBlock(block)
 	if err != nil {
 		log.Errorf("Could not hash received block: %v", err)
-		return [32]byte{}, false, err
+		return nil, nil, false, err
 	}
 
 	log.Debugf("Processing response to block request: %#x", blockRoot)
 	hasBlock := rs.db.HasBlock(blockRoot)
-	span.AddAttributes(trace.BoolAttribute("hasBlock", hasBlock))
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Revert this change

File: beacon-chain/sync/receive_block.go [hunk 6]
```diff
@@ -135,21 +140,19 @@ func (rs *RegularSync) validateAndProcessBlock(ctx context.Context, blockMsg p2p
...
+		log.Errorf("Could not update chain head: %v", err)
+		return nil, nil, false, err
 	}
 
 	sentBlocks.Inc()
 	// We update the last observed slot to the received canonical block's slot.
 	if block.Slot > rs.highestObservedSlot {
 		rs.highestObservedSlot = block.Slot
 	}
-	span.AddAttributes(trace.Int64Attribute("highestObservedSlot", int64(rs.highestObservedSlot)))
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Revert this change

File: beacon-chain/sync/receive_block.go [hunk 7]
```diff
@@ -88,36 +99,30 @@ func (rs *RegularSync) validateAndProcessBlock(ctx context.Context, blockMsg p2p
...
 
 	beaconState, err := rs.db.HeadState(ctx)
 	if err != nil {
 		log.Errorf("Failed to get beacon state: %v", err)
-		return [32]byte{}, false, err
+		return nil, nil, false, err
 	}
 
-	span.AddAttributes(
-		trace.Int64Attribute("block.Slot", int64(block.Slot)),
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: Revert this change.

Also, while you are here, can you change both of these values to be `x - params.BeaconConfig().GenesisSlot`? They overflow currently

File: beacon-chain/sync/receive_block.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/sync/receive_block.go]: Can we add an attribute to the span to say the block was invalid?

File: beacon-chain/sync/receive_block.go (hunk 1 above)
Comment (rauljordan) [reply beacon-chain/sync/receive_block.go]: Hey @prestonvanloon receiveBlock is no longer recursive, only processBlockAndFetchAncestors and that function does not call fork-choice.

Comment (prestonvanloon) [reply beacon-chain/sync/receive_block.go]: I'm wrong

File: beacon-chain/sync/receive_block.go (hunk 2 above)
Comment (rauljordan) [reply beacon-chain/sync/receive_block.go]: Added

File: beacon-chain/sync/receive_block.go [hunk 8]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+
```
Comment (terenc3t) [inline beacon-chain/sync/receive_block.go]:
```suggestion
```
